### ISSUE TITLE
How to change the base url at runtime?

### ISSUE THREAD

**Issue by zubairehman** [flutterchina/dio#402]
### Issue Description and Steps

Is there any way to change the base url at runtime. I have created singleton Dio object with base url (parent server url) to get client specific configurations from parent server. Once configurations are fetched, i need to change the base url and point to client server.

Dio Client:
```
  /// A singleton dio provider.
  ///
  /// Calling it multiple times will return the same instance.
  @provide
  @singleton
  Dio provideDio(SharedPreferenceHelper sharedPrefHelper) {
    return Dio()
      ..options.baseUrl = Endpoints.baseUrl
      ..options.connectTimeout = Endpoints.connectionTimeout
      ..options.receiveTimeout = Endpoints.receiveTimeout
      ..options.headers = {
        Headers.CONTENT_TYPE: 'application/json; charset=utf-8',
        Headers.ACCEPT: 'application/json',
        Headers.GARAGE_ID: Headers.TGP_GARAGE_ID
        //special header for TGP garage identification
      }
      ..interceptors.add(LogInterceptor(request: true, responseBody: true))
      ..interceptors.add(
        InterceptorsWrapper(
          onRequest: (Options options) async {
            
            // getting token from shared pref
            var _authToken = sharedPrefHelper.authToken;

            if (_authToken != null) {
              options.headers.putIfAbsent('Authorization', () => _authToken);
            }

          },
          onResponse: (Response response) {},
          onError: (DioError err) {

            // check to see if response code is 401, then either refresh the token
            // or take the user to login screen
            if (err != null &&
                err.response != null &&
                err.response.statusCode == 401) {
              print('Status is 401, please logout');

              // firing logout event, this will be caught on navigation widget
              // and it will clear all the data and will take user to login screen
              eventBus.fire(UserLoggedInEvent(false));
            }
          },
        ),
      );
  }
```

**Comment by Triipaxx** [comment]
Try creating a new Dio object with new base URL and credentials

**Comment by trevorwang** [comment]
#385

**Comment by wendux** [comment]
Fixed at v2.1.14.
